### ISSUE TITLE
Tpetra_ASSUME_CUDA_AWARE_MPI affecting test performance on white

### ISSUE THREAD

**Issue by kddevin** [trilinos/Trilinos#3405]
@trilinos/tpetra @trilinos/zoltan2 @ndellingwood @kyungjoo-kim @nmhamster @crtrott 

## Next Action Status

PR #3500 merged to 'develop' on 12/6/2018 that enables `Tpetra_ASSUME_CUDA_AWARE_MPI =ON` by default for ATDM (and all) CUDA Trilinos builds.  Next, watch to see ATDM Trilinos builds and the EMPIRE builds over next few days to see what happens ...

## Description

Using devpack/20180521/openmpi/3.1.0/gcc/7.2.0/cuda/9.2.88 on white, the behavior of tests differs depending on the setting of Tpetra_ASSUME_CUDA_AWARE_MPI.
With Tpetra_ASSUME_CUDA_AWARE_MPI=ON, there are segfaults in MPI_Send called from Tpetra::Distributor::doPosts.  
With Tpetra_ASSUME_CUDA_AWARE_MPI=OFF, the tests run fine.  (Note that ATDM testing uses this setting.)

I don't know whether this problem is caused by OpenMPI 3 (see #3356), or a problem with Tpetra, or something that I'm doing wrong with my configure/build.  @crtrott wrote the Tpetra test script that I am using (see info below), and it worked without errors in March with now obsolete devpack/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44.

I've had difficulty getting Trilinos to configure and build with devpacks using OpenMPI 2.
I will continue to try to get an OpenMPI 2 version to work.  I welcome an OpenMPI 2 build script that works on white; please share if you have one.  Thanks.

## Motivation and Context

We need Tpetra to work for Zoltan2 testing.
The segfaults from the Tpetra::Distributor's MPI_Send occur in Zoltan2 testing as well; setting Tpetra_ASSUME_CUDA_AWARE_MPI=OFF allows Zoltan2 tests to run.

## Steps to Reproduce

I am running the Tpetra test script described at https://github.com/trilinos/Trilinos/wiki/Tpetra-test-script.  I changed the devpack in the script to devpack/20180521/openmpi/3.1.0/gcc/7.2.0/cuda/9.2.88, because the original devpack in the script (devpack/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44) is no longer available on white.  I compared tests with -DTpetra_ASSUME_CUDA_AWARE_MPI=ON and OFF.  

On white, in my home directory, 
Trilinos/Obj_white/Test_white_2018_09_06_22.34.48 has tests with Tpetra_ASSUME_CUDA_AWARE_MPI=OFF, and 
Trilinos/Obj_white/Test_white_2018_09_06_15.35.24 has tests with Tpetra_ASSUME_CUDA_AWARE_MPI=ON.

The usual Testing/Temporary/LastTest.log shows the segfaults for the case with Tpetra_ASSUME_CUDA_AWARE_MPI=ON, and the passing tests with it OFF.

**Comment by mhoemmen** [comment]
Could this be the same problem with OpenMPI 3 that @nmhamster found on waterman?

btw if you want to experiment without rebuilding, just set the environment variable `TPETRA_ASSUME_CUDA_AWARE_MPI` to either 0 or 1.

**Comment by kyungjoo-kim** [comment]
Regarding to mpi2 build, I just tried it on white and encountered a configuration error as follows.
```
/home/projects/ppc64le-pwr8-nvidia/openmpi/2.1.2/gcc/7.2.0/cuda/9.0.176/bin/mpicxx   \
  -O3   -g  --std=c++11 -Xcudafe --diag_suppress=esa_on_defaulted_function_ignored \
-expt-extended-lambda -arch=sm_60 -Xcompiler -fopenmp blah blah... 
/home/projects/ppc64le/gcc/7.2.0/include/c++/7.2.0/bits/move.h(48): error: identifier "__builtin_addressof" is undefined  
```
This is because somehow the cuda compiler does not know about ``__builtin_addressof``. 

From googling (https://stackoverflow.com/questions/48639190/intel-icc-compile-c-code-cause-error), people suggest to put a patch to the move.h something like below.
```
  template<typename _Tp>                                                                                                                                                                       
    inline _GLIBCXX_CONSTEXPR _Tp*                                                                                                                                                             
    __addressof(_Tp& __r) _GLIBCXX_NOEXCEPT                                                                                                                                                    
#if defined(__CUDA_ARCH__) && probably need version check for 9.0 (9.2 is working)
    { return (&const_cast<char&>(reinterpret_cast<const volatile char&>(__r)));}
#else
    { return __builtin_addressof(__r); }  
#endif
```
Do we have cuda nightly test on mpi2 and cuda 9.0 with gcc 7.2 ? If so, this type of configuration error should be detected earlier. It is a bit weird that devpack is provided but the devpack is not tested with cmake.

**Comment by kddevin** [comment]
Thanks for trying this, @kyungjoo-kim 

With devpack/20180308/openmpi/2.1.2/gcc/7.2.0/cuda/9.0.176, I see the same problems as @kyungjoo-kim 

With the IBM compiler devpack/20180521/openmpi/2.1.2/ibm/xl/16.1.0/cuda/9.2.88
I get unresolved BLAS and LAPACK functions, even though the devpack loads netlib/3.8.0/ibm/xl/16.1.0 .

Thus, it has been impossible to test this build with OpenMPI 2 and CUDA.

@trilinos/framework @bartlettroscoe Any suggestions for an OpenMPI 2  CUDA build on white?  Thanks.

**Comment by bartlettroscoe** [comment]
From talking with @nmhamster last week, there are not supposed to be any problems the OpenMPI 3 on Power8 (i..e 'white' and 'ride'), just Power9 (i.e. 'waterman').

@kddevin, if you want to try the OpenMPI 2 env on 'white', I think you can just make a local branch and then modify the file `Trilinos/cmake/std/atdm/ride/environment.sh` and use the ATDM Trilinos configure process described at:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md#ridewhite

**Comment by kddevin** [comment]
Thanks for your reply, @bartlettroscoe 

I have already done exactly as you recommend.  (The ATDM process is nicely documented and easy enough to modify; thank you for that!)  I think the problem is with the devpacks on white; things are not quite consistent to build with CUDA and OpenMPI2.  I sent an email to white-help.  Fingers crossed.

**Comment by kddevin** [comment]
Yeah, I think this is a problem with OpenMPI3 on white.  It appears to be less CUDA-aware than OpenMPI2.  I emailed white-help.  Stay tuned.

One question would be why the ATDM tests don't set Tpetra_ASSUME_CUDA_AWARE_MPI=ON (and, thus, don't experience this problem)?  I see the following in the ATDM CMake output:
-- Determine whether Tpetra will assume that MPI is CUDA aware:
--   - Attempt to detect whether MPI is CUDA aware, by searching for "ompi_info" executable.
--   - Tpetra did not find the "ompi_info" executable.  This may not be bad; for example, if your MPI implementation is not OpenMPI, then you won't have this executable.  Tpetra will conservatively assume that your MPI implementation is NOT CUDA aware.  If you would like to change this, please set the CMake variable Tpetra_ASSUME_CUDA_AWARE_MPI:BOOL=ON explicitly at configure time.

In my build, I see
-- Determine whether Tpetra will assume that MPI is CUDA aware:
--   - Attempt to detect whether MPI is CUDA aware, by searching for "ompi_info" executable.
--   - Calling "ompi_info" (full path: /ascldap/users/projects/ppc64le-pwr8-nvidia/openmpi/3.1.0/gcc/7.2.0/cuda/9.2.88/bin/ompi_info) to find out whether MPI is CUDA aware.
--   - "ompi_info" explicitly claims that your MPI implementation is CUDA aware.

**Comment by mhoemmen** [comment]
@kddevin The only way to access OpenMPI's configuration information, is to run the `ompi_info` executable.  Look at `tpetra/CMakeLists.txt` to see that Tpetra goes through a lot of effort to find that executable in various directories.  If it can't find it, then Tpetra conservatively assumes by default that the MPI implementation is not CUDA aware.

**Comment by mhoemmen** [comment]
@kddevin What I mean is that it's bad that Tpetra can't find the `ompi_info` executable.  That means:

  - white is set up as a cross-compilation build (meaning that Tpetra won't try to run executables), or 
  - the MPI paths are messed up, or 
  - `ompi_info` didn't get installed, or
  - Tpetra's CMake logic is wrong or incomplete somehow.

**Comment by kddevin** [comment]
Thanks, @mhoemmen .  We'll need @bartlettroscoe to address the questions about the ATDM builds.  Since using CUDA-aware MPI is preferred, he'll likely want to investigate what's happening in his environment.

**Comment by bartlettroscoe** [comment]
> We'll need @bartlettroscoe to address the questions about the ATDM builds. Since using CUDA-aware MPI is preferred, he'll likely want to investigate what's happening in his environment.

@kddevin, what aspect of the env?  Note, I don't control the env, I just load the modules that I am told to load.  If there is something wrong with the loaded env, then we need to work with the people that admin those envs.

**Comment by kddevin** [comment]
@bartlettroscoe It would be good to figure out why the ATDM builds do not enable Tpetra_ASSUME_CUDA_AWARE_MPI; I would think they would want that.  Let me do a little digging to determine whether the ATDM setting comes from Trilinos or from something in the ATDM modules.  I might need your help after that.  Thanks.

**Comment by bartlettroscoe** [comment]
>  It would be good to figure out why the ATDM builds do not enable Tpetra_ASSUME_CUDA_AWARE_MPI; I would think they would want that.

@kddevin, if Trilinos does not enable them automatically already, then we can turn them on manually.  But before we do that, we need to ensure that this works for our ATDM customer codes.

>  Let me do a little digging to determine whether the ATDM setting comes from Trilinos or from something in the ATDM modules. I might need your help after that. Thanks.

This should hopefully all be explained in the document:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md

That should describe all of the files and the info they contain.  Hopefully that is easy to follow and find things.

**Comment by bartlettroscoe** [comment]
CC: @fryeguy52, @kddevin, @mhoemmen 

FYI: As part of #3482, I am turning this option on and testing on 'waterman' with GCC 7.2.0 + OpenMPI 2.1.2 + CUDA 9.2 to see if this helps avoid some timeouts (or even failures) In MueLu and Panzer tests.

**Comment by mhoemmen** [comment]
Is this really a Tpetra issue, or is it an MPI configuration issue?  We should write some raw MPI + CUDA benchmarks to see.  I think Christian wrote one a while back for diagnosing UVM issues.

**Comment by bartlettroscoe** [comment]
CC: @rppawlo, @bathmatt, @micahahoward, @mhoemmen 

> I left this issue open only because I was surprised that ATDM's tests didn't see the OpenMPI3 problem. It turns out they do not enable CUDA-aware MPI, which I thought was strange. When I have some spare time, I may investigate why.

@kddevin, just to be clear, the ATDM Trilinos configuration is not setting `Tpetra_ASSUME_CUDA_AWARE_MPI=OFF`, it is just not setting it to `Tpetra_ASSUME_CUDA_AWARE_MPI=ON`.  These were the settings we got from the EMPIRE configuration of Trilinos several months ago (but they may have diverged in the meantime).  But note that the SPARC CUDA configuration of Trilinos does not appear to be explicitly setting `Tpetra_ASSUME_CUDA_AWARE_MPI=ON` either.  Therefore, it appears that unless Trilinos implicitly sets this to on at configure time, then no ATDM applications are using this option.  That seems strange if this is such an important option for performance on the GPU.

**Comment by kddevin** [comment]
There is logic in Trilinos' build system to determine whether the MPI is CUDA-Aware.  For my configuration on white, it decides that, yes, MPI is CUDA-aware.  For the ATDM configuration, it decides it isn't.  I hope to get some time to determine why the behavior is different.

**Comment by kddevin** [comment]
BTW:  To determine what SPARC is using, you can look at the CMake output as in my notes above from September 17.  They show the output from your build and mine.

**Comment by bartlettroscoe** [comment]
FYI: I occurred to me that perhaps the older GCC 5.4.0 + OpenMPI 1.10.4 + CUDA 8.0.44 env on 'white' and 'ride' were set up to have Tpetra automatically detect that MPI was CUDA aware.  But that appears **not** to be the case as shown in the last  `Trilinos-atdm-white-ride-cuda-debug` build on `white` before the upgrade on July 17 shown [here](https://testing.sandia.gov/cdash-dev-view/viewConfigure.php?buildid=3728145) which showed:

```
Processing enabled package: Tpetra (Classic, Core, Tests, Examples)
-- Determine whether Tpetra will assume that MPI is CUDA aware:
--   - Attempt to detect whether MPI is CUDA aware, by searching for "ompi_info" executable.
--   - Tpetra did not find the "ompi_info" executable.  This may not be bad; for example, if your MPI implementation is not OpenMPI, then you won't have this executable.  Tpetra will conservatively assume that your MPI implementation is NOT CUDA aware.  If you would like to change this, please set the CMake variable Tpetra_ASSUME_CUDA_AWARE_MPI:BOOL=ON explicitly at configure time.
```

So my guess is that on 'white' and 'ride' EMPIRE was never using MPI-aware features in Tpetra.

**Comment by mhoemmen** [comment]
@bartlettroscoe I think this could only happen if `ompi_info` didn't get installed into either `$PATH` or `MPI_BIN_DIR`, or if the build was set up to be a cross-compilation build.

**Comment by bartlettroscoe** [comment]
FYI: Setting `Tpetra_ASSUME_CUDA_AWARE_MPI=ON` on 'waterman' with the installed GCC 7.2.0 + OpenMPI 2.1.2 + CUDA 9.2 env just creates more test failures.  See https://github.com/trilinos/Trilinos/pull/3482#issuecomment-423753635.

**Comment by kddevin** [comment]
> @kddevin What I mean is that it's bad that Tpetra can't find the `ompi_info` executable. That means:
>     * white is set up as a cross-compilation build (meaning that Tpetra won't try to run executables), or
>     * the MPI paths are messed up, or
>     * `ompi_info` didn't get installed, or 
>     * Tpetra's CMake logic is wrong or incomplete somehow.

@mhoemmen ; this is a good clue; thanks.
The ATDM instructions prescribe building on the login node and submitting batch jobs only to run the tests.  The tests I was running did the configure and build through the batch system.   I'm going with your "cross-compilation" theory.

**Comment by bartlettroscoe** [comment]
> The ATDM instructions prescribe building on the login node and submitting batch jobs only to run the tests. The tests I was running did the configure and build through the batch system. I'm going with your "cross-compilation" theory.

@kddevin, we need to update the instructions for these specific machines, but for all of the test bed machines, you can configure and build on either the login node or a compute node.  They should have about identical environments.  The nightly builds all build on the compute nodes (which you can see evidence of that [here](https://testing.sandia.gov/cdash-dev-view/viewConfigure.php?buildid=3968206),  for example, which shows `Trilinos_HOSTNAME='white25' which is a compute node) so I don't think there is any cross-compilation issue here.  Where is `ompi_info` supposed to exist?

**Comment by mhoemmen** [comment]
@bartlettroscoe It should be a normal part of OpenMPI's installation.

**Comment by bartlettroscoe** [comment]
> It should be a normal part of OpenMPI's installation, and should live in the same binary directory as all the other OpenMPI executables.

@mhoemmen, @kddevin 

Looks like `ompi_info` is there.  On the 'waterman' on the login node:

```
$ ssh waterman

$ hostname
waterman11

$ cd ~/Trilinos.base/Trilinos

$ . cmake/std/atdm/load-env.sh cuda-9.2-debug
Hostname 'waterman11' matches known ATDM host 'waterman' and system 'waterman'
ATDM_CONFIG_TRILNOS_DIR = /home/rabartl/Trilinos.base/Trilinos
Setting default compiler and build options for JOB_NAME='cuda-9.2-debug'
No KOKKOS_ARCH specified so using system default
Using waterman compiler stack CUDA-9.2 to build DEBUG code with Kokkos node type CUDA and KOKKOS_ARCH=Power9,Volta70

$ module list
Currently Loaded Modulefiles:
  1) git/2.10.1                                              12) zlib/1.2.8
  2) binutils/2.30.0                                         13) hdf5/1.8.20/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
  3) gcc/7.2.0                                               14) cgns/20180517/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
  4) cuda/9.2.88                                             15) matio/1.5.11/gcc/7.2.0
  5) java/ibm/sdk/8.0.0                                      16) pnetcdf-exo/1.9.0/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
  6) numa/2.0.11                                             17) netcdf-exo/4.6.1/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
  7) openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88                     18) parmetis/4.0.3/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
  8) cmake/3.6.2                                             19) boost/1.65.1/gcc/7.2.0
  9) valgrind/3.12.0                                         20) superlu-dist/4.3.0/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
 10) netlib/3.8.0/gcc/7.2.0                                  21) devpack/20180517/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
 11) metis/5.0.1/gcc/7.2.0

$ which mpiexec
/home/projects/ppc64le-pwr9-nvidia/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88/bin/mpiexec

$ which ompi_info
/home/projects/ppc64le-pwr9-nvidia/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88/bin/ompi_info

$ ompi_info --version
Open MPI v2.1.2

http://www.open-mpi.org/community/help/
```

And to test this on the compute node, I used the script `get_ompi_info.sh`:

```
#!/bin/bash

set -x
hostname
set +x

source ~/Trilinos.base/Trilinos/cmake/std/atdm/load-env.sh cuda-9.2-debug

set -x

module list

which ompi_info

ompi_info --version
```

and when I ran this on a compute node with:

```
$  bsub -x -Is -n 20 ./get_ompi_info.sh
```

it returned:

```
***Forced exclusive execution
Job <29095> is submitted to default queue <rhel7W>.
<<Waiting for dispatch ...>>
<<Starting on waterman2>>
+ hostname
waterman2
+ set +x
Hostname 'waterman2' matches known ATDM host 'waterman' and system 'waterman'
ATDM_CONFIG_TRILNOS_DIR = /home/rabartl/Trilinos.base/Trilinos
Setting default compiler and build options for JOB_NAME='cuda-9.2-debug'
No KOKKOS_ARCH specified so using system default
Using waterman compiler stack CUDA-9.2 to build DEBUG code with Kokkos node type CUDA and KOKKOS_ARCH=Power9,Volta70
+ module list
++ /usr/bin/modulecmd bash list
Currently Loaded Modulefiles:
  1) git/2.10.1                                              12) zlib/1.2.8
  2) binutils/2.30.0                                         13) hdf5/1.8.20/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
  3) gcc/7.2.0                                               14) cgns/20180517/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
  4) cuda/9.2.88                                             15) matio/1.5.11/gcc/7.2.0
  5) java/ibm/sdk/8.0.0                                      16) pnetcdf-exo/1.9.0/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
  6) numa/2.0.11                                             17) netcdf-exo/4.6.1/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
  7) openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88                     18) parmetis/4.0.3/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
  8) cmake/3.6.2                                             19) boost/1.65.1/gcc/7.2.0
  9) valgrind/3.12.0                                         20) superlu-dist/4.3.0/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
 10) netlib/3.8.0/gcc/7.2.0                                  21) devpack/20180517/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
 11) metis/5.0.1/gcc/7.2.0
+ eval
+ which ompi_info
/home/projects/ppc64le-pwr9-nvidia/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88/bin/ompi_info
+ ompi_info --version
Open MPI v2.1.2

http://www.open-mpi.org/community/help/
```

So there must be some type of defect in CMake or the CMakeLists.txt code.

Who can debug this?

**Comment by mhoemmen** [comment]
@bartlettroscoe I can help, just not today very much -- I need to finish something for another project.  I'm curious if the formatting of `ompi_info` output changed.

**Comment by kddevin** [comment]
@mhoemmen I have both successful and unsuccessful builds already working on white; let me take a quick look and then I'll hit you with questions.

**Comment by mhoemmen** [comment]
FYI, to help people answer Ross' question, it's possible to enable this feature at run time, by setting an environment variable, even if it was disabled by default at configure time.  The main value to setting this configure-time option is to enable the feature by default for tests.

**Comment by kddevin** [comment]
@mhoemmen  The code should run faster with CUDA-aware MPI, right?  
If so, we should encourage the ATDM testing and ATDM users to use it.

**Comment by mhoemmen** [comment]
> The code should run faster with CUDA-aware MPI, right?

We hope it would run faster.  If not, then there may be something wrong with our code.

**Comment by kddevin** [comment]
Thanks, @mhoemmen 
@bartlettroscoe Given the expected benefits to performance, I think it is worth trying to make this option work in the ATDM builds.  Perhaps you could try them with the CUDA-aware option ON in the ATDM environment and let us know the test results.   Then we can assess what is needed to fix any failing tests.  Thanks!

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe Given the expected benefits to performance, I think it is worth trying to make this option work in the ATDM builds. Perhaps you could try them with the CUDA-aware option ON in the ATDM environment and let us know the test results. Then we can assess what is needed to fix any failing tests. Thanks!

@kddevin, I am running the CUDA builds right now on 'white'.  See https://github.com/trilinos/Trilinos/pull/3500#issuecomment-438808627.  If that goes well, we might just cross our fingers and merge this.  Then if there are problems on other systems we can deal with them (either by backing out the merge and fixing offline out or fixing them in short order).

**Comment by bartlettroscoe** [comment]
PR #3500 merged to 'develop' on 12/6/2018 that enables `Tpetra_ASSUME_CUDA_AWARE_MPI =ON` by default for ATDM (and all) CUDA Trilinos builds.  Next, watch to see ATDM Trilinos builds over next few days to see what happens ...

I added the "in review" flag.  Hopefully we can get some results on 'waterman' over the next few days to see what impact this has.  (Or perhaps my job will actually run that I started in #3500 to confirm this.)
